libs/vm-modules/tests/unit/ml_model_tests.cpp

```diff
@@ -32,11 +32,52 @@ namespace {
 
 using DataType = fetch::vm_modules::math::DataType;
 
+static std::string const ADD_INVALID_LAYER_TEST_SOURCE = R"(
+    function main()
+      var model = Model("sequential");
+      <<TOKEN>>
+    endfunction
+  )";
+
+static char const *INVALID_LAYER_TYPE =
+    R"(model.add("INVALID_LAYER_TYPE", 1u64, 1u64, 1u64, 1u64);)";
+
+static char const *INVALID_ACTIVATION_DENSE =
+    R"(model.add("dense", 10u64, 10u64, "INVALID_ACTIVATION_DENSE");)";
+
+static char const *INVALID_ACTIVATION_CONV =
+    R"(model.add("conv1d", 1u64, 1u64, 1u64, 1u64, "INVALID_ACTIVATION_CONV");)";
+
+static char const *INVALID_PARAMS_DENSE_NOACT = R"(model.add("dense", 1u64, 1u64, 1u64, 1u64);)";
+static char const *INVALID_PARAMS_DENSE_RELU =
+    R"(model.add("dense", 1u64, 1u64, 1u64, 1u64, "relu");)";
+
+static char const *INVALID_PARAMS_CONV_NOACT = R"(model.add("conv1d", 10u64, 10u64);)";
+static char const *INVALID_PARAMS_CONV_RELU  = R"(model.add("conv1d", 10u64, 10u64, "relu");)";
+
+static char const *WRONG_PARAM_COUNT_CONV  = R"(model.add("conv1d", 10u64, 10u64, 10u64, "relu");)";
+static char const *WRONG_PARAM_COUNT_DENSE = R"(model.add("dense", 10u64, 10u64, 10u64, "relu");)";
+static char const *WRONG_PARAM_TYPE_CONV =
+    R"(model.add("conv1d", 0u64, 10fp32, 10u64, 10u64, "relu");)";
+static char const *WRONG_PARAM_TYPE_DENSE = R"(model.add("dense", 10fp32, 10u64, "relu");)";
+
 class VMModelTests : public ::testing::Test
 {
 public:
   std::stringstream stdout;
   VmTestToolkit     toolkit{&stdout};
+
+  void TestInvalidLayerCompilation(std::vector<std::string> sources)
+  {
+    for (const auto &test_case : sources)
```

Comment:
still you could ditch this for-loop, take in just a single source, and separate all the tests for the most clarity

Reply:
Ok, I'll do it